cmd/block-generator/runner/run.go

```diff
@@ -0,0 +1,260 @@
+			return fmt.Errorf("unable to parse metric '%s': %w", metric, err)
+		}
+	}
+	rate := sum / count
+
+	msg := fmt.Sprintf("%s:%f\n", key, rate)
+	if _, err := out.WriteString(msg); err != nil {
+		return fmt.Errorf("unable to write metric '%s': %w", key, err)
+	}
+	return nil
+}
+
+// Run the test for 'RunDuration', collect metrics and write them to the 'ReportDirectory'
+func (r *Args) runTest(indexerURL string, generatorURL string) error {
+	collector := &MetricsCollector{MetricsURL: fmt.Sprintf("http://%s/metrics", indexerURL)}
+
+	baseName := filepath.Base(r.Path)
+	baseNameNoExt := strings.TrimSuffix(baseName, filepath.Ext(baseName))
+	reportPath := path.Join(r.ReportDirectory, fmt.Sprintf("%s.report", baseNameNoExt))
+
+	report, err := os.Create(reportPath)
+	if err != nil {
+		return fmt.Errorf("unable to create report: %w", err)
+	}
+	defer report.Close()
+
+	// Run for r.RunDuration
+	start := time.Now()
+	for time.Since(start) < r.RunDuration {
+		time.Sleep(r.RunDuration / 10)
+		if err := collector.Collect("indexer_daemon_import_time_sec"); err != nil {
+			return fmt.Errorf("problem collecting metrics: %w", err)
+		}
+	}
+	if err := collector.Collect("indexer_daemon_import_time_sec"); err != nil {
+		return fmt.Errorf("problem collecting metrics: %w", err)
+	}
+
+	// Collect results.
+	durationStr := fmt.Sprintf("test_duration_seconds:%d\n", uint64(r.RunDuration.Seconds()))
+	if _, err := report.WriteString(durationStr); err != nil {
+		return fmt.Errorf("unable to write duration metric: %w", err)
+	}
+
+	resp, err := http.Get(fmt.Sprintf("http://%s/report", generatorURL))
+	if err != nil {
+		return fmt.Errorf("the process failed to start properly, health endpoint query failed")
+	}
+	defer resp.Body.Close()
+	var generatorReport generator.Report
+	if err := json.NewDecoder(resp.Body).Decode(&generatorReport); err != nil {
+		return fmt.Errorf("problem decoding generator report: %w", err)
+	}
+	for metric, entry := range generatorReport {
+		// Skip this one
+		if metric == "genesis" {
+			continue
+		}
+		str := fmt.Sprintf("transaction_%s_total:%d\n", metric, entry.GenerationCount)
+		if _, err := report.WriteString(str); err != nil {
+			return fmt.Errorf("unable to write transaction_count metric: %w", err)
+		}
+	}
+
+	// Record a rate from one of the first data points.
+	if len(collector.Data) > 5 {
+		if err := recordDataToFile(collector.Data[2], "starting_block_import_duration_average_seconds", report); err != nil {
+			return err
+		}
+	}
+
+	// Also record the final one.
+	if err := recordDataToFile(collector.Data[len(collector.Data)-1], "final_import_duration_average_second", report); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// startGenerator starts the generator server.
+func startGenerator(configFile string, addr string) func() error {
+	// Start generator.
+	server, done := generator.StartServer(configFile, addr)
+
+	return func() error {
+		if err := server.Shutdown(context.Background()); err != nil {
```

Comment:
From the description, `Shutdown()` blocks until the server is shut down and `ListenAndServe()` returns immediately. So, `done` will be closed immediately. On a related note, I think we shouldn't worry about shutting down the generator gracefully.

Reply:
Both these functions block. I happen to be reusing the same port for each run, so it's important that the generator and indexer both shut down before starting the next test.

Reply:
Yes, this should exit immediately if everything is working correctly. If it doesn't there's a 10 second timeout to make sure nothing hangs.

Reply:
Fixed, thanks!